### PR TITLE
CCGLView: Add createWithFullscreen overloaded method that sets video mode.

### PR DESCRIPTION
**Pull request by wefiends**
There is currently no way to pass the desired size for a fullscreen window. It always sets the window size to the current video mode.

This adds an overloaded method that passes a size to GLFW. If the size doesn't match an available video mode, GLFW will set it to closest available size.

### CODE REVIEW

File: cocos/2d/platform/desktop/CCGLView.cpp [hunk 1]
```diff
@@ -374,6 +386,15 @@ bool GLView::initWithFullScreen(const std::string& viewName)
     return initWithRect(viewName, Rect(0, 0, videoMode->width, videoMode->height), 1.0f);
 }
 
+bool GLView::initWithFullScreen(const std::string &viewName, cocos2d::Size size, float frameZoomFactor)
```
Comment (dumganhar) [inline cocos/2d/platform/desktop/CCGLView.cpp]: Why does full screen need `size`?

Comment (dumganhar) [inline cocos/2d/platform/desktop/CCGLView.cpp]: And frameZoomFactor isn't needed since you never use it.

Comment (wefiends) [inline cocos/2d/platform/desktop/CCGLView.cpp]: When creating a fullscreen window with GLFW, setting the size will cause GLFW to switch video modes.

Comment (wefiends) [inline cocos/2d/platform/desktop/CCGLView.cpp]: I'll drop the frameZoomFactor param and pass in 1.0f.

Comment (dumganhar) [inline cocos/2d/platform/desktop/CCGLView.cpp]: But how does user use this function? What `size` should it pass in?
Is it possible to add a API like:

``` c++
GLView::createWithFullScreen(const std::string& viewname, int videoMode)
```

?

Comment (wefiends) [inline cocos/2d/platform/desktop/CCGLView.cpp]: I can change the method to take a `const GLFWvidmode` instead and create a `Rect` to pass to `initWithRect`.

 Developers need to retrieve the available video modes:
`GLFWmonitor *monitor = glfwGetPrimaryMonitor();`
`int count = 0;`
`const GLFWvidmode *modes = glfwGetVideoModes(monitor, &count);`

And choose one that's appropriate for their game. I can add a method that returns that array in CCGLView.

Ultimately, GLFW expects a width and height when creating the window, and it chooses the closest available video mode. You cannot directly pass it a `GLFWvidMode`.

Comment (dumganhar) [inline cocos/2d/platform/desktop/CCGLView.cpp]: > Ultimately, GLFW expects a width and height when creating the window, and it chooses the closest available video mode. You cannot directly pass it a GLFWvidMode.

good to know this. :)

**Comment by dumganhar** [comment]
When i tested this PR on mac which has two monitors, whatever the `Size` value is, the game view is always at my main monitor (MBP's screen).
PS: I used MBP 2013 retina(2880*1800) with an external screen with 1920x1080 resolution.

**Comment by wefiends** [comment]
Yes, the current createWithFullscreen defaults to the primary monitor (as reported by the OS). This was the behavior even before my patch.

In the latest commit:

`createWithFullScreen(const std::string& viewName, const GLFWvidmode &videoMode, GLFWmonitor *monitor)`
Create the fullscreen window on a specific monitor, setting resolution, refresh rate, and rgb bit depth. Monitor and video mode should be retrieved via GLFW at runtime.

We're using this method in our game to allow user resolution switching. We save their settings and retrieve the video mode and monitor at startup.

**Comment by ricardoquesada** [comment]
+1 for supporting this feature (although I haven't read the patch).

Linux, Mac and Windows games are important for us, so +1 for adding features that are needed for this.
One feature that we need is a way to switch from fullscreen to window mode in runtime. I think GLFW 3.1 will support it.

File: cocos/2d/platform/desktop/CCGLView.h [hunk 2]
```diff
@@ -39,6 +39,7 @@ class CC_DLL GLView : public GLViewProtocol, public Ref
     static GLView* create(const std::string& viewName);
     static GLView* createWithRect(const std::string& viewName, Rect size, float frameZoomFactor = 1.0f);
     static GLView* createWithFullScreen(const std::string& viewName);
+    static GLView* createWithFullScreen(const std::string& viewName, Size size);
```
Comment (dumganhar) [inline cocos/2d/platform/desktop/CCGLView.h]: It seems that you forgot to update `CCGLView.h`.

Comment (dumganhar) [inline cocos/2d/platform/desktop/CCGLView.h]: This `createWithFullScreen` has been removed, right?

Comment (wefiends) [inline cocos/2d/platform/desktop/CCGLView.h]: Yes, sorry. It's removed in the latest commit.

Comment (dumganhar) [inline cocos/2d/platform/desktop/CCGLView.h]: Thank you. I will check it some hours later. :)

**Comment by dumganhar** [comment]
Thanks.

**Pull request merged**